Introduce standalone build

gergondet opened the pull request:
The CI runs the simulation in both settings (catkin/standalone) to confirm the proper working of both setups.

@mmurooka can you double check f9ce9741a5123ebf473226346dd2bc163931272f this seems more coherent with the cnoid project but maybe you kept the value for actual experiments?

gergondet committed:
Introduce standalone build

gergondet committed:
Change the topic for the cart pose

gergondet committed:
Implement both catkin and standalone workflow

Check simulation results in both cases for one of the build
configuration

mmurooka commented:
Thank you for updates! 

> can you double check https://github.com/isri-aist/LocomanipController/commit/f9ce9741a5123ebf473226346dd2bc163931272f this seems more coherent with the cnoid project but maybe you kept the value for actual experiments?

(I haven't seen the details of the changes yet, so my reply below may not be to the point.)
For CI simulations, the suggested change is not necessary as long as the following configuration is correctly loaded. I think `/object/pose` looks more generic than `/cnoid/object/pose` (assuming real environment and MuJoCo), so it is appropriate for the default, but I have no strong preference. However, if the CI simulations do not work without this change, that is not the intention.

https://github.com/isri-aist/LocomanipController/blob/e7798561383da4dbba47235c882e95948f4b2257/.github/workflows/config/PushCartVelMode.yaml#L20-L21

https://github.com/isri-aist/LocomanipController/blob/e7798561383da4dbba47235c882e95948f4b2257/.github/workflows/config/PushCartWaypoint.yaml#L28-L29

These are loaded by

https://github.com/isri-aist/LocomanipController/blob/e7798561383da4dbba47235c882e95948f4b2257/.github/workflows/ci-catkin.yaml#L129

I will definitely remove the configuration overwrite before the controller constructor that we have been discussing (sorry for late), but I believe the overwrite via `~/.config/mc_rtc/controllers/<ControllerName>.yaml` is still acceptable.

mmurooka commented on .github/workflows/ci.yaml:
```diff
@@ -30,85 +30,69 @@ jobs:
             mc-rtc-version: head
+            catkin-build: catkin
+            motion-type: PushCartVelMode
+          - os: ubuntu-20.04
```
This new `include` is not needed. Instead, delete the `catkin-build: catkin` line from the original `include` above. (If I understand the syntax correctly)

gergondet replied:
I had this in the first iteration and it only generates one configuration without any `catkin-build` defined (see https://github.com/gergondet/LocomanipController/actions/runs/4301782536/jobs/7499440259 and https://github.com/gergondet/LocomanipController/actions/runs/4301782536/workflow).

My understanding of the [doc](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations) is that `include` will create new build configuration if it overwrites an existing key in the configuration and this does not include other values in the matrix (the `- fruit: banana` example in the docuentation)

mmurooka replied:
I see, thanks for the information. :ok_hand:

mmurooka commented on .github/workflows/ci.yaml:
```diff
@@ -30,85 +30,69 @@ jobs:
+        if: matrix.catkin-build == 'standalone'
+        uses: jrl-umi3218/github-actions/install-dependencies@master
+        with:
+          compiler: ${{ matrix.compiler }}
```
`matrix.compiler` is not defined, but is it OK?

mmurooka commented on .github/workflows/ci.yaml:
```diff
@@ -30,85 +30,69 @@ jobs:
+          ubuntu: |
+            apt: libforcecontrolcollection-dev libtrajectorycollection-dev libcentroidalcontrolcollection-dev libmc-rtc-dev
+          github: |
+            - path: isri-aist/BaselineFootstepPlanner
```
```suggestion
```
`BaselineFootstepPlanner` is optional dependency for `BaselineWalkingController` and it can be skipped for this CI.

mmurooka commented on .github/workflows/ci.yaml:
```diff
@@ -30,85 +30,69 @@ jobs:
+        with:
+          compiler: ${{ matrix.compiler }}
+          build-type: ${{ matrix.build-type }}
+          ros: |
+            apt: sbpl
```
```suggestion
```
`sbpl` is not necessary when `BaselineFootstepPlanner` is deleted as in the comment below.

mmurooka commented on .github/workflows/ci.yaml:
```diff
@@ -118,23 +102,37 @@ jobs:
-          cp ${GITHUB_WORKSPACE}/catkin_ws/src/isri-aist/LocomanipController/.github/workflows/config/${{ matrix.motion-type }}.yaml ${HOME}/.config/mc_rtc/controllers/LocomanipController.yaml
+          if [ "${{ matrix.catkin-build }}" == "catkin" ]
+          then
+            cp ${GITHUB_WORKSPACE}/catkin_ws/src/${{ github.repository }}/etc/mc_rtc.yaml ${HOME}/.config/mc_rtc
```
```suggestion
            cp ${GITHUB_WORKSPACE}/catkin_ws/src/isri-aist/LocomanipController/etc/mc_rtc.yaml ${HOME}/.config/mc_rtc
```

In the following `depends.rosinstall`, it is described as `isri-aist`, so we should not use `${{ github.repository }}`, right?
https://github.com/isri-aist/LocomanipController/blob/e7798561383da4dbba47235c882e95948f4b2257/depends.rosinstall#L2

gergondet replied:
This is getting the `mc_rtc.yaml` generated by `LocomanipController` which is not described in `depends.rosinstall` and is cloned by the `build-catkin-project` action in `${GITHUB_WORKSPACE}/catkin_ws/src/${{ github.repository }}` ([here](https://github.com/jrl-umi3218/github-actions/blob/master/build-catkin-project/action.yml#L37))

mmurooka replied:
Sorry, it's OK as it is.

gergondet committed:
Address review comments

gergondet commented:
Thanks for the review. I have addressed your comments in [`9e7b2db` (#3)](https://github.com/isri-aist/LocomanipController/pull/3/commits/9e7b2db7e507fa24a0e38050dbdec36e5f341615)

> For CI simulations, the suggested change is not necessary as long as the following configuration is correctly loaded. I think /object/pose looks more generic than /cnoid/object/pose (assuming real environment and MuJoCo), so it is appropriate for the default, but I have no strong preference. However, if the CI simulations do not work without this change, that is not the intention.

I had not realized the overwrite was done in the CI configuration files. I did the "fix" to get it to work in simulation on my machine but as an "example" configuration it makes sense not to have the `/cnoid` prefix. So I have reverted the change in my last commit.

mmurooka commented on CMakeLists.txt:
```diff
@@ -1,46 +1,67 @@
+  set(BWC_STATES_FILES "${BWC_STATES_LIBRARIES}/data")
+  set(STATES_INSTALL_PREFIX "${CATKIN_DEVEL_PREFIX}/lib")
+  set(CONFIG_OUT "${CATKIN_DEVEL_PREFIX}/lib/mc_controller/etc/LocomanipController.yaml")
+  configure_file(etc/LocomanipController.in.yaml ${CONFIG_OUT})
```
This line is duplicated.
```suggestion
```

mmurooka reviewed (approved):
OK to merge if https://github.com/isri-aist/LocomanipController/pull/3#discussion_r1123977869 is resolved. Thanks.

gergondet committed:
Remove duplicate configure_file

gergondet commented:
Thanks. I have removed the duplicate line and will merge when the CI passes
